pkg/controller/manager.go

```diff
@@ -81,33 +81,56 @@ func (m *Manager) updateController(name string, params ControllerParams) *manage
 
 		ctrl.getLogger().Debug("Controller update time: ", time.Since(start))
 	} else {
-		ctrl = &managedController{
-			controller: controller{
-				name:       name,
-				group:      params.Group,
-				uuid:       uuid.New().String(),
-				stop:       make(chan struct{}),
-				update:     make(chan ControllerParams, 1),
-				trigger:    make(chan struct{}, 1),
-				terminated: make(chan struct{}),
-			},
-		}
-		ctrl.updateParamsLocked(params)
-		ctrl.getLogger().Debug("Starting new controller")
-
-		m.controllers[ctrl.name] = ctrl
-		m.mutex.Unlock()
+		return m.createControllerAndUnlock(name, params)
+	}
 
-		globalStatus.mutex.Lock()
-		globalStatus.controllers[ctrl.uuid] = ctrl
-		globalStatus.mutex.Unlock()
+	return ctrl
+}
 
-		go ctrl.runController(ctrl.params)
+func (m *Manager) createControllerAndUnlock(name string, params ControllerParams) *managedController {
+	ctrl := &managedController{
+		controller: controller{
+			name:       name,
+			group:      params.Group,
+			uuid:       uuid.New().String(),
+			stop:       make(chan struct{}),
+			update:     make(chan ControllerParams, 1),
+			trigger:    make(chan struct{}, 1),
+			terminated: make(chan struct{}),
+		},
 	}
+	ctrl.updateParamsLocked(params)
+	ctrl.getLogger().Debug("Starting new controller")
 
+	m.controllers[ctrl.name] = ctrl
+	m.mutex.Unlock()
+
+	globalStatus.mutex.Lock()
+	globalStatus.controllers[ctrl.uuid] = ctrl
+	globalStatus.mutex.Unlock()
+
+	go ctrl.runController(ctrl.params)
 	return ctrl
 }
 
+// CreateController installs a new controller in the
+// manager.  If a controller with the name already exists
+// this method returns false without triggering, otherwise
+// creates the controller and runs it immediately.
+func (m *Manager) CreateController(name string, params ControllerParams) bool {
+	m.mutex.Lock()
+	if m.controllers != nil {
+		if _, exists := m.controllers[name]; exists {
+			m.mutex.Unlock()
+			return false
+		}
+	} else {
+		m.controllers = controllerMap{}
+	}
+	m.createControllerAndUnlock(name, params)
```

Comment:
Nit: I think I would prefer a `createControllerLocked` and just do the unlocking of the mutex at the caller level, rather than mixing responsibilities.

Reply:
Didn't want to change the locking semantics (e.g. to hold both this manager and the globalStatus manager locks). But looks like we do that in e.g. `RemoveController` anyway. I'll change it to `createControllerLocked`.